underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsPositionedUnderFileInputStream.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs.hdfs;
+
+import alluxio.underfs.SeekableUnderFileInputStream;
+import alluxio.util.io.BufferUtils;
+
+import com.google.common.base.Preconditions;
+import org.apache.hadoop.fs.FSDataInputStream;
+
+import java.io.IOException;
+
+/**
+ * The input stream of HDFS as under filesystem. This input stream supports seeking but internally
+ * uses the positionedRead {@link FSDataInputStream} API. This stream can be cached for reuse.
+ */
+public class HdfsPositionedUnderFileInputStream extends SeekableUnderFileInputStream {
+  private long mPos;
+
+  HdfsPositionedUnderFileInputStream(FSDataInputStream in, long pos) {
+    super(in);
+    mPos = pos;
+  }
+
+  @Override
+  public int available() throws IOException {
+    int available = in.available() + (int) (((FSDataInputStream) in).getPos() - mPos);
+    return available > 0 ? available : 0;
+  }
+
+  @Override
+  public long getPos() throws IOException {
+    return mPos;
+  }
+
+  @Override
+  public int read() throws IOException {
+    byte[] buffer = new byte[1];
+    int bytesRead = ((FSDataInputStream) in).read(mPos, buffer, 0, buffer.length);
+    if (bytesRead > 0) {
+      mPos += bytesRead;
+      return BufferUtils.byteToInt(buffer[0]);
+    }
+    Preconditions.checkArgument(bytesRead != 0,
+        "Expected a non-zero value if end of stream has not been reached");
```

Comment:
Is it valid to hit EOS here and return `-1`?

Reply:
yes, `-1` is valid return value for this method